core/src/main/scala/org/apache/spark/util/SignalUtils.scala

```diff
@@ -58,18 +58,45 @@ private[spark] object SignalUtils extends Logging {
    */
   def register(signal: String)(action: => Boolean): Unit = synchronized {
     if (SystemUtils.IS_OS_UNIX) {
-      try {
-        val handler = handlers.getOrElseUpdate(signal, {
-          logInfo("Registering signal handler for " + signal)
-          new ActionHandler(new Signal(signal))
-        })
-        handler.register(action)
-      } catch {
-        case ex: Exception => logWarning(s"Failed to register signal handler for " + signal, ex)
-      }
+      register(signal, s"Failed to register signal handler for " + signal,
+        logStackTrace = true)(action)
     }
   }
 
+  /**
+   * Adds an action to be run when a given signal is received by this process.
+   *
+   * This method receives failMessage as additional parameter, which would be logged when it fails
+   * to register the signal. Here the failures include the cases 1) OS doesn't support signal at
+   * all 2) OS doesn't support given signal (Could be possible with non-POSIX signals)
+   *
+   * All actions for a given signal are run in a separate thread.
+   */
+  def register(
+      signal: String,
+      failMessage: => String,
```

Comment:
I suppose this can just be `String` rather than delay evaluation, but it doesn't matter much.

Reply:
I was intended to defer evaluation of building message like the way we do with logXXX, though this patch doesn't leverage it. As it doesn't matter much for now, I'll simply change it.